### ISSUE TITLE
MFM support for DFS disks

### ISSUE THREAD

**Issue by eelcohuininga** [bmcollier/beebem#12]
Solidisk not only made controllers, but also DFS'es. In the current version of BeebEm (v4.14) there is only support for MFM discs for ADFS, not for DFS.

**Comment by jgharston** [comment]
eelcohuininga wrote:
> Solidisk not only made controllers, but also DFS'es. In the current
> version of BeebEm (v4.14) there is only support for MFM discs for
> ADFS, not for DFS.

Isn't that hardware specific not filing system specific? What happens if
you do:
Machine=Master
*TAPE (ie neither DFS /or/ ADFS)
?drivecontrol=DDEN+SIDE0+DS0
read from floppy disk controller

On the Electron and Compact* I can certainly access double-density disks
regardless of the selected filing system. If I NOP out the start of my
*FORMAT command I wrote last month so it doesn't do IF FS=ADFS THEN...
it certainly works regardless of selected filing system. After all, it's
just doing hardware access, nothing at all to do with the filing system.

*My Master's PSU is currently dead.

-- 
J.G.Harston - jgh@mdfs.net - mdfs.net/jgh

**Comment by eelcohuininga** [comment]
No, it is not filing system specific. When BeebEm loads a disc image it determines the type of disc based upon the 'Total No. Of Sectors' stored in sector 1, bytes 6 and 7. Since BeebEm v4.14 only recognizes total sector sizes of &190 and &320, it will never recognize DFS disks with size &280 or &500, and therefore output an error message that the disk format isn't recognized. Also, the code for recognizing the disk format was kinda messy, and I've cleaned it up a bit.

I do wonder how Watford DFS behaves when loading DD/MFM disks. I never had experience with this type of DFS so this would be a nice test case...

On a completely different note: both the disk8271.c and disk1770.c modules contain disk loading/new disk procedures. Ideally these procedures should be FDC independant, and the disk8271.c and disk1770.c modules should be restricted to just emulating the FDC. Now, when the load-disk needs to be updated, both procedures in both .c files should be updated, which is inefficient and confusing. But that's another issue, not yet reported on this GitHub site...

**Comment by irrelevantdotcom** [comment]
I would agree with that latter point. There were several disc upgrades available which were all incompatible, particularly before Acorn brought out the 1770, (I know, I had one...) and it would make it easier to support those if the FDC emulation code was divorced from generic routines.

**Comment by eelcohuininga** [comment]
Ok, I will open a new issue about the latter point, where we can discuss if and how the updating of disk8271.c and disk1770.c should be realized.

**Comment by jgharston** [comment]
eelcohuininga wrote:
> No, it is not filing system specific. When BeebEm loads a disc image
> it determines the type of disc based upon the 'Total No. Of Sectors'
> stored in sector 1, bytes 6 and 7. Since BeebEm v4.14 only recognizes
> total sector sizes of &190 and &320, it will never recognize DFS disks
> with size &280 or &500, (...)

You should really do it on the image size. Well, really, you should just
let whatever is accessing the emulated hardware access it and let it 
decide
instead of having the emulator impose its own opinion on the image file.

Otherwise, doing that means that some AcornCPM disks will be rejected as
they will be "seen" to only have two sectors as the disk size in the
catalogue is set to put the CP/M contents after the logical end of the 
disk.

> I do wonder how Watford DFS behaves when loading DD/MFM disks. I never
> had experience with this type of DFS so this would be a nice test
> case...

Initially it does a 'scan sector IDs' to find the sector headers. If
accessing the disk results in a format mis-match, it swaps the density
and tries again.

> On a completely different note: both the disk8271.c and disk1770.c
> modules contain disk loading/new disk procedures. Ideally these
> procedures should be FDC independant, and the disk8271.c and
> disk1770.c modules should be restricted to just emulating the FDC.

Agreed.

-- 
J.G.Harston - jgh@mdfs.net - mdfs.net/jgh

**Comment by eelcohuininga** [comment]
Op 9-2-2016 om 21:00 schreef jgharston:
> eelcohuininga wrote:
> > No, it is not filing system specific. When BeebEm loads a disc image
> > it determines the type of disc based upon the 'Total No. Of Sectors'
> > stored in sector 1, bytes 6 and 7. Since BeebEm v4.14 only recognizes
> > total sector sizes of &190 and &320, it will never recognize DFS disks
> > with size &280 or &500, (...)
>
> You should really do it on the image size. Well, really, you should just
> let whatever is accessing the emulated hardware access it and let it
> decide
> instead of having the emulator impose its own opinion on the image file.

I would, but I haven't written that part of BeebEm :-)

I totally agree with you, but the way BeebEm is coded right now is that 
it decides which disk type is loaded by examining the Total Sector Size, 
and not on hardware features like track/sector size, encoding techniques 
etc.  So, without rewriting all 8271 and 1770-code, this was the 
quickest way for me to implement MFM support for DFS.

Ideally BeebEm should /internally /use raw disk images which /exactly 
/mimics the hardware (with gap bytes, sector id's etc.).  That way the 
emulator just presents a raw disk image to the FDC-emulator code, which 
in turn just presents bytes to the FS-rom. It should resemble the 
original hardware as close as possible imo.

It would not just mean implementing raw image support as I suggested in 
issue #8, but a total rewrite of the code. If everyone agrees we could 
open an issue stating something like "BeebEm uses raw disk images 
internally" or something likewise...

Cheers
